phoenix-core/src/main/java/org/apache/phoenix/iterate/RegionScannerFactory.java

```diff
@@ -205,6 +210,22 @@ public boolean nextRaw(List<Cell> result) throws IOException {
              */
             IndexUtil.wrapResultUsingOffset(env, result, offset, dataColumns,
                 tupleProjector, dataRegion, indexMaintainer, viewConstants, ptr);
+
+            byte[] expBytes = scan.getAttribute(BaseScannerRegionObserver.LOCAL_INDEX_FILTER);
```

Comment:
This is really the crux of the change.
Get the WHERE expression from the scan and evaluate it *after* we have assembled the tuple. This cannot be done with just a regular filter since that is evaluated before.

Reply:
Does the client issue a new scan for every row to be returned? Otherwise, how does this work if there are multiple rows to be returned to the client for a query? It seems that it is assumed here that the order of rows to be visited in the index CF is the same as the order of of rows to be visited in the data CF (based on how scanTillScanStartRow is used). If so, this assumption is wrong. Maybe I misunderstood the implementation here.

Reply:
Thanks  for looking @kadirozde .

No, there's only a single scan. Just like filters the expression here remains the same for the entire scan.

The intention here is that every every row that makes through the index scan (that's where start/end row and scanTillScanStartRow come into the picture) is subsequently passed through to the expression to see whether it should be filtered by non-index condition. The key is that is has be done after we merged in the columns from the main column family.

Also note that the filter expression here only contains the none index-filters (see WhereCompiler.setScanFilter(...) and how that is called).

Reply:
The scan scans in index order. The other non-index columns are then merged in via an in-region Get to the main column family.

Reply:
As @lhofhansl pointed out on a private discussion, the scan initiated by the client is used to visit the index CF and then the data table rows are picked up using Get operations within IndexUtil.wrapResultUsingOffset.